integrations/tensorflow/e2e/iree_e2e_product_test_suite.bzl

```diff
@@ -0,0 +1,229 @@
+        to take a cartesian product of.
+      tags:
+        tags to apply to the test. Note that as in standard test suites, manual
+        is treated specially and will also apply to the test suite itself.
+      data:
+        external data for iree_py_test.
+      deps:
+        test dependencies for iree_py_test.
+      size:
+        size of the tests for iree_py_test.
+      python_version:
+        the python version to run the tests with. Uses python3 by default.
+      **kwargs:
+        any additional arguments that will be passed to the underlying tests and
+        test_suite.
+    """
+
+    # Normalize flags_to_values to always have lists as its values.
+    # e.g. {use_external_data: True} -> {use_external_data: [True]}
+    flags_to_values = normalize_dictionary(flags_to_values)
+
+    all_flag_configurations = dictionary_product(flags_to_values)
+
+    failing_flag_configurations = []
+    if failing_configurations != None:
+        for failing_configuration in failing_configurations:
+            failing_configuration = normalize_dictionary(failing_configuration)
+
+            # If a flag isn't specified in the failing configuration, assume it
+            # is failing for all values of that flag.
+            for key, value in flags_to_values.items():
+                if key not in failing_configuration:
+                    failing_configuration[key] = value
+
+            failing_flag_configurations.extend(
+                dictionary_product(failing_configuration),
+            )
+
+    tests = []
+    for flags in all_flag_configurations:
+        # Check if this is a failing configuration.
+        failing = flags in failing_flag_configurations
+
+        # Append "_failing" to name if this is a failing configuration.
+        test_name = name if not failing else name + "_failing"
+        test_name = [test_name]
+        for k, v in flags.items():
+            # Only include the flag's value in the test name if it's not always
+            # the same.
+            if len(flags_to_values[k]) > 1:
+                test_name.append(k)
+                test_name.append(str(v))
+        test_name = "__".join(test_name)
+        tests.append(test_name)
+
+        args = ["--{}={}".format(k, v) for k, v in flags.items()]
+
+        py_test_tags = []
+        if "target_backends" in flags:
+            # TODO(GH-2175): Simplify this after backend names are standardized.
+            # "iree_<driver>" --> "<driver>"
+            backend = flags["target_backends"]
+            if len(backend.split(",")) > 1:
+                fail("Expected only one target backend but got '{}'".format(
+                    backend,
+                ))
+
+            driver = backend.replace("iree_", "")
+            if driver == "llvmjit":
+                driver = "llvm"
+            py_test_tags += ["driver={}".format(driver)]
+            if tags != None:  # `is` is not supported.
+                py_test_tags += tags
```

Comment:
This special treatment is at odds with this very general rule.

Reply:
I was a bit conflicted here. On one hand the macro is written as if you could avoid passing `target_backends`, but on the other this macro is explicitly supposed to be used within our e2e integration tests. A few thoughts:

- I could make `target_backends` a kwarg on the macro, but I would rather not because it would be out of place relative to all of the other flags, and I would rather be consistent about how flags are handled.
- I could de-duplicate this code and write a function that is called by both `iree_e2e_test_suite` and this macro.
- I could make including `target_backends` in `flags_to_values` mandatory, since excluding it would be a bug.

Reply:
Doing the last 2 sounds like an improvement to me

Reply:
Done :)